remill/BC/DeadStoreEliminator.cpp

```diff
@@ -0,0 +1,768 @@
+      auto aamd = offset_to_aamd[offset];
+      load_inst->setAAMetadata(aamd);
+    } else if (auto store_inst = llvm::dyn_cast<llvm::StoreInst>(inst)) {
+      auto aamd = offset_to_aamd[offset];
+      store_inst->setAAMetadata(aamd);
+    }
+  }
+}
+
+// Return a map of MDNode scopes and a vector of AAMDNodes based on the given vector of StateSlots,
+// where each byte offset (i.e. index) in the slots vector is mapped to a
+// corresponding AAMDNodes struct.
+AAMDInfo::AAMDInfo(const std::vector<StateSlot> &slots, llvm::LLVMContext &context) {
+  // Create a vector of pairs of scopes to slot offsets
+  // This will be made into a map at the end of the function
+  // We need it as a vector for now so that it is ordered when creating the noalias
+  std::vector<std::pair<llvm::MDNode *, uint64_t>> scope_offsets;
+  scope_offsets.reserve(slots.size());
+  for (const auto &slot : slots) {
+    auto mdstr = llvm::MDString::get(context, "slot_" + std::to_string(slot.i));
+    scope_offsets.emplace_back(llvm::MDNode::get(context, mdstr), slot.offset);
+  }
+  std::vector<llvm::AAMDNodes> aamds;
+  // One AAMDNodes struct for each byte offset so that we can easily connect them
+  aamds.reserve(slots.size());
+  for (uint64_t i = 0; i < slots.size(); i++) {
+    if (aamds.empty() || slots[i].i != slots[i - 1].i) {
+    // noalias all slots != scope
+    std::vector<llvm::Metadata *> noalias_vec;
+    noalias_vec.reserve(slots.back().i + 1);
+    for (uint64_t j = 0; j < slots.size(); j++) {
+      if (i != j && (noalias_vec.empty() || scope_offsets[j].first != noalias_vec.back())) {
+        noalias_vec.push_back(scope_offsets[j].first);
+      }
+    }
+    llvm::MDNode *noalias = llvm::MDNode::get(context, llvm::MDTuple::get(context, noalias_vec));
+    aamds.emplace_back(nullptr, scope_offsets[i].first, noalias);
+    } else {
+      // copy the last element
+      aamds.push_back(aamds.back());
+    }
+  }
+  std::unordered_map<llvm::MDNode *, uint64_t> scopes(scope_offsets.begin(), scope_offsets.end());
+  slot_scopes = scopes;
+  slot_aamds = aamds;
+}
+
+void AnalyzeAliases(llvm::Module *module, const std::vector<StateSlot> &slots) {
+  const AAMDInfo aamd_info(slots, module->getContext());
+  llvm::DataLayout dl = module->getDataLayout();
+  auto bb_func = BasicBlockFunction(module);
+  for (auto &func : *module) {
+    if (&func != bb_func && !func.isDeclaration() &&
+        func.getFunctionType() == bb_func->getFunctionType()) {
+      auto sp = LoadStatePointer(&func);
+      ForwardAliasVisitor fav(slots, &dl, sp);
+      LOG(INFO) << "Analyzing aliases of func " << func.getName().str();
+      for (auto &block : func) {
+        for (auto &inst : block) {
+          fav.AddInstruction(&inst);
+        }
+      }
+      // if the analysis succeeds for this function, add the AAMDNodes
+      if (fav.Analyze()) {
+        LOG(INFO) << "Offsets: " << fav.offset_map.size();
+        LOG(INFO) << "Aliases: " << fav.alias_map.size();
```

Comment:
Let's rename `offset_map` and `alias_map` to something like `state_offset` and `state_access_offset`, respectively.